src/services/loopd/loop-service.ts

```diff
@@ -0,0 +1,255 @@
+
+  // helpers
+  const clientHealthCheck = util.promisify<TokensRequest, TokensResponse>(
+    swapClient.getLsatTokens.bind(swapClient),
+  )
+  const clientSwapOut = util.promisify<LoopOutRequest, SwapResponse>(
+    swapClient.loopOut.bind(swapClient),
+  )
+  const clientSwapOutQuote = util.promisify<QuoteRequest, OutQuoteResponse>(
+    swapClient.loopOutQuote.bind(swapClient),
+  )
+  const clientSwapOutTerms = util.promisify<TermsRequest, OutTermsResponse>(
+    swapClient.loopOutTerms.bind(swapClient),
+  )
+
+  const healthCheck = async (): Promise<boolean | SwapServiceError> => {
+    try {
+      const request = new TokensRequest()
+      const resp = await clientHealthCheck(request)
+      if (resp) return true
+    } catch (error) {
+      return new SwapErrorHealthCheckFailed(error)
+    }
+    return false
+  }
+
+  const swapOut = async function ({
+    amount,
+    maxSwapFee,
+    swapDestAddress,
+  }: SwapOutArgs): Promise<SwapOutResult | SwapServiceError> {
+    const fee = maxSwapFee ? maxSwapFee : 20000
+    try {
+      const request = new LoopOutRequest()
+      // on regtest, set the publication deadline to 0 for faster swaps, otherwise
+      // set it to 30 minutes in the future to reduce swap fees
+      const thirtyMins = 30 * 60 * 1000
+      const swapPublicationDeadline =
+        btcNetwork === BtcNetwork.regtest ? 0 : Date.now() + thirtyMins
+      if (swapDestAddress) request.setDest(swapDestAddress)
+      request.setAmt(amount)
+      request.setMaxSwapFee(fee)
+      request.setMaxPrepayRoutingFee(fee)
+      request.setMaxSwapFee(fee)
+      request.setMaxPrepayAmt(fee)
+      request.setMaxMinerFee(fee)
+      request.setSwapPublicationDeadline(swapPublicationDeadline)
+      request.setInitiator(`galoy-${loopdInstanceName}`)
+      if (btcNetwork === BtcNetwork.regtest) request.setHtlcConfirmations(1)
+      if (btcNetwork === BtcNetwork.regtest) request.setSweepConfTarget(2)
+      const resp = await clientSwapOut(request)
+      const swapOutResult: SwapOutResult = {
+        htlcAddress: resp.getHtlcAddress(),
+        serverMessage: resp.getServerMessage(),
+        swapId: resp.getId(),
+        swapIdBytes: resp.getIdBytes().toString(),
+      }
+      return swapOutResult
+    } catch (error) {
+      if (error.message.includes("channel balance too low")) {
+        return new SwapErrorChannelBalanceTooLow(error)
+      }
+      return new SwapServiceError(error)
+    }
+  }
+
+  const swapListener = function (): SwapClientReadableStream<SwapListenerResponse> {
+    try {
+      const request = new MonitorRequest()
+      const listener = swapClient.monitor(request)
+      // listener.on("data" | "status" | "end")
+      listener.on("data", (data: SwapStatus & SwapStatusResultWrapper) => {
+        listener.pause()
+        // parse data to our interface
+        const stateVal = data.getState()
+        let state
+        try {
+          state = Object.keys(SwapState).find((key) => {
```

Comment:
I'm having a hard time following the logic

Reply:
I moved the parsing logic out to separate functions. it's a bit confusing because the auto generated proto code uses numbers as the values and the human readable name as the key (see code below) . so you need to parse using `Object.keys` and then map to our Swap Domain model

```typescript
export interface FailureReasonMap {
  FAILURE_REASON_NONE: 0;
  FAILURE_REASON_OFFCHAIN: 1;
  FAILURE_REASON_TIMEOUT: 2;
  FAILURE_REASON_SWEEP_TIMEOUT: 3;
  FAILURE_REASON_INSUFFICIENT_VALUE: 4;
  FAILURE_REASON_TEMPORARY: 5;
  FAILURE_REASON_INCORRECT_AMOUNT: 6;
}
```